sdk/src/main/java/io/opentelemetry/sdk/metrics/export/IntervalMetricReader.java

```diff
@@ -0,0 +1,181 @@
+
+/**
+ * Wraps a list of {@link MetricProducer}s and automatically reads and exports the metrics every
+ * export interval.
+ *
+ * @since 0.3.0
+ */
+public final class IntervalMetricReader {
+  private final Exporter exporter;
+  private final ScheduledFuture<?> scheduledFuture;
+
+  /**
+   * Stops the worker thread by calling {@link Thread#interrupt()}.
+   *
+   * @since 0.3.0
+   */
+  public void shutdown() {
+    scheduledFuture.cancel(false);
+    exporter.run();
+  }
+
+  /**
+   * Returns a new {@link Builder} for {@link IntervalMetricReader}.
+   *
+   * @return a new {@link Builder} for {@link IntervalMetricReader}.
+   */
+  public static Builder builder() {
+    return new Builder(InternalState.builder());
+  }
+
+  /**
+   * Builder for {@link IntervalMetricReader}.
+   *
+   * @since 0.3.0
+   */
+  public static final class Builder {
+    private final InternalState.Builder optionsBuilder;
+
+    private Builder(InternalState.Builder optionsBuilder) {
+      this.optionsBuilder = optionsBuilder;
+    }
+
+    /**
+     * Sets the export interval.
+     *
+     * @param exportIntervalMillis the export interval between pushes to the exporter.
+     * @return this.
+     * @since 0.3.0
+     */
+    public Builder setExportIntervalMillis(long exportIntervalMillis) {
+      optionsBuilder.setExportIntervalMillis(exportIntervalMillis);
+      return this;
+    }
+
+    /**
+     * Sets the exporter to be called when export metrics.
+     *
+     * @param metricExporter the {@link MetricExporter} to be called when export metrics.
+     * @return this.
+     * @since 0.3.0
+     */
+    public Builder setMetricExporter(MetricExporter metricExporter) {
+      optionsBuilder.setMetricExporter(metricExporter);
+      return this;
+    }
+
+    /**
+     * Sets a collection of {@link MetricProducer} from where the metrics should be read.
+     *
+     * @param metricProducers a collection of {@link MetricProducer} from where the metrics should
+     *     be read.
+     * @return this.
+     * @since 0.3.0
+     */
+    public Builder setMetricProducers(Collection<MetricProducer> metricProducers) {
+      optionsBuilder.setMetricProducers(metricProducers);
+      return this;
+    }
+
+    /**
+     * Builds a new {@link IntervalMetricReader} with current settings.
+     *
+     * @return a {@code IntervalMetricReader}.
+     * @since 0.3.0
+     */
+    public IntervalMetricReader build() {
+      InternalState internalState = optionsBuilder.build();
+      Utils.checkArgument(
+          internalState.getExportIntervalMillis() > 0, "Export interval must be positive");
+
+      return new IntervalMetricReader(internalState);
+    }
+  }
+
+  private IntervalMetricReader(InternalState internalState) {
+    this.exporter = new Exporter(internalState);
+    ScheduledExecutorService scheduler =
+        Executors.newScheduledThreadPool(1, MoreExecutors.platformThreadFactory());
+    this.scheduledFuture =
```

Comment:
I'd never seen this pattern before. A more typical pattern in Java is to keep a reference to the ExecutorService and shut it down, rather than cancelling the future. I think if you don't explicitly shut down the executor service, it can block JVM exit in some scenarios.

Reply:
Done.